src/libre/re_ast.c

```diff
@@ -0,0 +1,389 @@
+	res->t = AST_EXPR_OPT;
+	res->u.opt.e = e;
+	LOG("-- %s: %p <- %p\n", __func__, (void *)res, (void *)e);
+	return res;
+}
+
+struct ast_expr *
+re_ast_expr_with_count(struct ast_expr *e, struct ast_count count)
+{
+	struct ast_expr *res = NULL;
+	if (count.low > count.high) {
+		fprintf(stderr, "ERROR: low > high (%u, %u)\n",
+		    count.low, count.high);
+		abort();
+	}
+
+	if (count.low == 0) {
+		if (count.high == 0) {		         /* 0,0 -> empty */
+			return re_ast_expr_empty();
+		} else if (count.high == 1) {		 /* 0,1 -> ? */
+			/* This isn't strictly necessary, but avoids
+			 * some unnecessary states and epsilon
+			 * transitions. */
+			return re_ast_opt(e);
+		} else if (count.high == AST_COUNT_UNBOUNDED) { /* 0,_ -> * */
+			return re_ast_kleene(e);
+		}
+	} else if (count.low == 1) {
+		if (count.high == AST_COUNT_UNBOUNDED) { /* 1,_ -> + */
+			return re_ast_plus(e);
+		} else if (count.high == 1) {	         /* 1,1 -> as-is */
+			return e;
+		}
+	}
+			
+	res = calloc(1, sizeof(*res));
+	if (res == NULL) { return res; }
+	res->t = AST_EXPR_REPEATED;
+	res->u.repeated.e = e;
+	res->u.repeated.low = count.low;
+	res->u.repeated.high = count.high;
+	LOG("-- %s: %p <- %p (%u,%u)\n", __func__, (void *)res,
+	    (void *)e, count.low, count.high);
+	return res;
+}
+
+struct ast_expr *
+re_ast_expr_char_class(struct re_char_class_ast *cca,
+    const struct ast_pos *start, const struct ast_pos *end)
+{
+	struct ast_expr *res = calloc(1, sizeof(*res));
+	if (res == NULL) { return res; }
+
+	res->t = AST_EXPR_CHAR_CLASS;
+	res->u.char_class.cca = cca;
+	memcpy(&res->u.char_class.start, start, sizeof *start);
+	memcpy(&res->u.char_class.end, end, sizeof *end);
+	LOG("-- %s: %p <- %p\n", __func__, (void *)res, (void *)cca);
+	return res;
+}
+
+struct ast_expr *
+re_ast_expr_char_type(enum ast_char_type_id id)
+{
+	struct ast_expr *res = calloc(1, sizeof(*res));
+	if (res == NULL) { return res; }
+
+	res->t = AST_EXPR_CHAR_TYPE;
+	res->u.char_type.id = id;
+	LOG("-- %s: %p <- %d\n", __func__, (void *)res, id);
+	return res;
+}
+
+struct ast_expr *
+re_ast_expr_group(struct ast_expr *e)
+{
+	struct ast_expr *res = calloc(1, sizeof(*res));
+	if (res == NULL) { return res; }
+	res->t = AST_EXPR_GROUP;
+	res->u.group.e = e;
+	res->u.group.id = (unsigned)-1; /* not yet assigned */
+	LOG("-- %s: %p <- %p\n", __func__, (void *)res, (void *)e);
+	return res;
+}
+
+struct ast_expr *
+re_ast_expr_re_flags(enum re_flags pos, enum re_flags neg)
+{
+	struct ast_expr *res = calloc(1, sizeof(*res));
+	if (res == NULL) { return res; }
+	res->t = AST_EXPR_FLAGS;
+	res->u.flags.pos = pos;
+	res->u.flags.neg = neg;
+	LOG("-- %s: %p <- pos %04x, neg %04x\n", __func__, (void *)res,
+	    (unsigned)pos, (unsigned)neg);
+	return res;
+}
+
+
+
+#define INDENT(F, IND)							\
+	do {								\
+		size_t i;						\
+		for (i = 0; i < IND; i++) { fprintf(F, " "); }		\
+	} while(0) 							\
+
+static void
+pp_iter(FILE *f, size_t indent, struct ast_expr *n)
+{
```

Comment:
Could you move the AST (and character class) pretty-printing to its own file please? At some point we can pull that out to a subdirectory and provide various output formats. I'm after dot and some kind of BNF especially.

Reply:
Moved in ae2f91d.